frontend/src/pages/dashboard/EntitySelector.tsx

```diff
@@ -123,24 +182,55 @@ export function CompoundSelector({ className }: SelectorProps): React.ReactEleme
           }}
         />,
       ]}
+      optionsFilter={compoundsFilter}
     />
   );
 }
 
-export function TargetSelector({ className }: SelectorProps): React.ReactElement {
-  const compoundsLoadable = useRecoilValueLoadable(targetsQuery);
+export function TargetSelector({ className, experiments }: SelectorProps): React.ReactElement {
+  const filterNeeds = useRecoilValueLoadable(
+    waitForAll([compoundSelectedQuery, projectSelectedQuery])
+  );
+  const targetsLoadable = useRecoilValueLoadable(targetsQuery);
   const targetSelected = useRecoilValue(targetSelectedQuery);
   const setTargetSelected = useSetRecoilState(targetIdSelectedAtom);
+
+  const targetsFilter = useCallback(
+    (targets: TargetDto[]) => {
+      if (filterNeeds.state === "hasValue" && filterNeeds.contents && experiments) {
+        const [selectedCompound, selectedProject] = filterNeeds.contents;
+        const targetNames = experiments
+          .filter(
+            experiment =>
+              (isDefined(selectedProject)
+                ? experiment.projectName === selectedProject.item.name
+                : true) &&
+              (isDefined(selectedCompound)
+                ? experiment.compoundName === selectedCompound.item.name
+                : true)
+          )
+          .map(experiment => experiment.targetName);
+
+        return targets.filter(
+          target => !!targetNames.find(targetName => targetName === target.item.name)
+        );
+      }
+      return undefined;
+    },
+    [filterNeeds, experiments]
+  );
```

Comment:
I am not a TS expert, but it looks like substantial copy-pasted. Did you try to define some helper to decrease code duplication?

Reply:
Hm, ok, it's changed in the next commit and apparently doesn't look like too massive copy-paste now.

Reply:
still looks like decent amount of copy-paste to me